### PR TITLE
Test defence organisation model with required and optional fields

### PR DESCRIPTION
**Pull request by laurentqro**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/LASB-575)

For each serialiser in CDA, test that it can serialise a resource when it has only the required fields, and when it has all the fields.

## Why

CDA frequently sees an entire class of bugs that are due to the fact that our serialisers assume an optional field is always present.

### CODE REVIEW

File: spec/models/defence_organisation_spec.rb [hunk 1]
```diff
@@ -1,18 +1,16 @@
 # frozen_string_literal: true
 
 RSpec.describe DefenceOrganisation, type: :model do
-  subject(:defendant) { described_class.new(body: defence_organisation_hash) }
-
-  let(:defence_organisation_hash) do
-    JSON.parse(file_fixture("defence_organisation.json").read)
+  subject(:defendant) do
```
Comment (arthurashman) [inline spec/models/defence_organisation_spec.rb:4]: Would this be clearer as `defence_organisation`?

File: spec/serializer/defence_organisation_serializer_spec.rb [hunk 2]
```diff
@@ -3,27 +3,32 @@
...
-                    name: "The Johnson Partnership",
-                    address1: "104",
-                    address2: "Fleet Street",
-                    address3: "Westminster",
-                    address4: "London",
-                    address5: "GB",
-                    postcode: "EC4A 2AH")
+  context "with required fields" do
+    let(:defence_organisation_data) { JSON.parse(file_fixture("defence_organisation/required_fields.json").read) }
+    let(:defence_organisation) { DefenceOrganisation.new(body: defence_organisation_data) }
```
Comment (arthurashman) [inline spec/serializer/defence_organisation_serializer_spec.rb:8]: My preference would be to have a fake DefenceOrganisation here so we are testing the serializer in isolation of the model. What are your thoughts?

File: spec/serializer/defence_organisation_serializer_spec.rb [hunk 3]
```diff
@@ -3,27 +3,32 @@
...
+    it { expect(attributes[:address4]).to be_nil }
+    it { expect(attributes[:address5]).to be_nil }
+    it { expect(attributes[:postcode]).to be_nil }
   end
 
-  context "with attributes" do
-    let(:attribute_hash) { subject[:data][:attributes] }
+  context "with all fields" do
+    let(:defence_organisation_data) { JSON.parse(file_fixture("defence_organisation/all_fields.json").read) }
+    let(:defence_organisation) { DefenceOrganisation.new(body: defence_organisation_data) }
```
Comment (arthurashman) [inline spec/serializer/defence_organisation_serializer_spec.rb:23]: Same here obviously

**Review by arthurashman — DISMISSED** [review]

File: spec/serializer/defence_organisation_serializer_spec.rb (hunk 2 above)
Comment (laurentqro) [reply spec/serializer/defence_organisation_serializer_spec.rb:8]: Yeah good point. And then we can use the fixtures to test the model itself.

File: spec/serializer/defence_organisation_serializer_spec.rb [hunk 4]
```diff
@@ -15,15 +15,13 @@
                     postcode: "EC4A 2AH")
   end
 
-  context "with attributes" do
-    let(:attribute_hash) { subject[:data][:attributes] }
+  let(:attributes) { subject[:data][:attributes] }
```
Comment (arthurashman) [inline spec/serializer/defence_organisation_serializer_spec.rb:19]: Sorry to be a pedant, but just for completion, could we have 
`it { expect(attributes[:id]).to eq("UUID") }`

Or even 11223344 if you want to change the double too

File: spec/models/defence_organisation_spec.rb [hunk 5]
```diff
@@ -1,18 +1,29 @@
...
-  it { expect(defendant.name).to eq("SOME ORGANISATION") }
-  it { expect(defendant.address1).to eq("102") }
-  it { expect(defendant.address2).to eq("Petty France") }
-  it { expect(defendant.address3).to eq("Floor 5") }
-  it { expect(defendant.address4).to eq("St James") }
-  it { expect(defendant.address5).to eq("Westminster") }
-  it { expect(defendant.postcode).to eq("EC4A 2AH") }
+  context "with all fields" do
+    let(:defence_organisation_data) { JSON.parse(file_fixture("defence_organisation/all_fields.json").read) }
+
```
Comment (arthurashman) [inline spec/models/defence_organisation_spec.rb:20]: `it { expect(defence_organisation.name.id).to eq("11223344") }`

File: spec/models/defence_organisation_spec.rb [hunk 6]
```diff
@@ -1,18 +1,29 @@
...
 
 RSpec.describe DefenceOrganisation, type: :model do
-  subject(:defendant) { described_class.new(body: defence_organisation_hash) }
+  subject(:defence_organisation) { described_class.new(body: defence_organisation_data) }
 
-  let(:defence_organisation_hash) do
-    JSON.parse(file_fixture("defence_organisation.json").read)
+  context "with required fields" do
+    let(:defence_organisation_data) { JSON.parse(file_fixture("defence_organisation/required_fields.json").read) }
+
```
Comment (arthurashman) [inline spec/models/defence_organisation_spec.rb:8]: `it { expect(defence_organisation.name.id).to be_nil }`

**Review by arthurashman — CHANGES_REQUESTED** [review]

File: spec/models/defence_organisation_spec.rb (hunk 6 above)
Comment (laurentqro) [reply spec/models/defence_organisation_spec.rb:8]: another good spot, thanks!

**Review by arthurashman — APPROVED** [review]

**Pull request merged**